### ISSUE TITLE
CPU affinity setting not working

### ISSUE THREAD

**Issue by smith558** [nfstream/nfstream#159]
I believe this may not be the correct wat to set CPU affinity on Linux systems.
https://github.com/nfstream/nfstream/blob/9751500db638e683a27d242fd4a308a4740e15c2/nfstream/utils.py#L138-L147

See
![image](https://user-images.githubusercontent.com/44023416/218440435-15dc5f8e-74a7-4d8b-910a-87b994e22cae.png)

https://github.com/giampaolo/psutil/issues/956#issuecomment-276059368

**Comment by smith558** [comment]
You should be respecting any existing configured affinity of the parent process. Perhaps with something like:
https://github.com/smith558/nfstream/blob/aacaf9d4641b6dfd2c2bc95281998db1f3150f29/nfstream/utils.py#L138-L155

**Comment by smith558** [comment]
Related to https://github.com/nfstream/nfstream/issues/149

**Comment by smith558** [comment]
Will open a PR for my proposal fix.

**Comment by aouinizied** [comment]
@smith558 Good catch! Are you planning to do a PR?

Thanks a lot!

**Comment by smith558** [comment]
> @smith558 Good catch! Are you planning to do a PR?
> 
> Thanks a lot!

I actually was. I was about to open one but then found out via experimentation there's an inherent issue in my solution that I seem to be unable to solve myself.

**Comment by aouinizied** [comment]
@smith558 Can you share your approach code snippert and more context about the issue you found. I can try to investigate it too.

**Comment by smith558** [comment]
@aouinizied The `psutil.Process().cpu_affinity(cpus)` call from the streamer seems to set the affinity for the whole Python process. This means if there's one Python script creating multiple NFStream objects (fairly expected use case), the subsequent object creations after the 1st one will only see the CPUs from `psutil.Process().cpu_affinity()` that was assigned by the very last `set_affinity(idx)` call. At least this seems to be the behaviour I am experiencing. This breaks multi-threading as the subsequent NFSteam streamers and meters all use the same threads. I had to be storing all affinity-respecting CPUs available upon the object creation because it was breaking even the allocation for meters within the first NFStream object (however, after the "fix" proposed below that is not necessary).

A visual example below from an actual experiment on an 8-core 2-thread (16 CPUS) machine (the first process below after object creation is always the streamer and the following two are the meters)
```
NFStreamer objects created sequentially
NFStreamer 1 (n_meters=2)
CPUs available:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
setting process to: [0, 1]
CPUs available:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
setting process to: [2, 3]
CPUs available:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
setting process to: [4, 5]
NFStreamer 2 (n_meters=2)
CPUs available:  [0, 1]
setting process to: [0, 1]
CPUs available:  [0, 1]
setting process to: [0, 1]
CPUs available:  [0, 1]
setting process to: [0, 1]
```
I just found out this behaviour is caused by:
https://github.com/nfstream/nfstream/blob/7cdb20163b1c09e5b16f3985a38859c03b52ccb9/nfstream/streamer.py#L346
Is this call necessary here? What is the purpose of it? Removing this line fixes the issue.
```
NFStreamer objects created sequentially
NFStreamer 1 (n_meters=3)
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
CPUs available:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
setting process to: [2, 3]
NFStreamer 2 (n_meters=3)
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
CPUs available:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
setting process to: [2, 3]
CPUs available:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
setting process to: [4, 5]
CPUs available:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
setting process to: [6, 7]
```

**Comment by smith558** [comment]
I am happy to open the PR for the current state so that we have the code changes I made visible at least.

**Comment by aouinizied** [comment]
@smith558 Yes go ahead with a PR and I will merge it. Note that you can remove streamer pinning causing the issue. Your approach is safer. 

Zied

**Comment by smith558** [comment]
> @smith558 Yes go ahead with a PR and I will merge it. Note that you can remove streamer pinning causing the issue. Your approach is safer.
> 
> Zied

The only issue with that is that then we lose control over which CPU the streamer gets assigned to. Is that okay?

**Comment by aouinizied** [comment]
@smith558 You can do it this way or you can send back e message from last meter that will trigger affinity set on the parent process.

**Comment by aouinizied** [comment]
@smith558 I implemented a fix in https://github.com/nfstream/nfstream/commit/0634f7ac0300bce47dbaeeb23a87b0157cabab0e 

Can you have a look and confirm that it solve your issue?

Thanks,
Zied

**Comment by smith558** [comment]
> @smith558 I implemented a fix in [0634f7a](https://github.com/nfstream/nfstream/commit/0634f7ac0300bce47dbaeeb23a87b0157cabab0e)
> 
> Can you have a look and confirm that it solve your issue?
> 
> Thanks, Zied

Yes, I can try running this version a few times to see if it works, the code looks like it should.

**Comment by smith558** [comment]
@aouinizied  I am not exactly sure what this new function is doing (no comments and complicated logic) but it seems to be having undesirable effects.
https://github.com/nfstream/nfstream/blob/0634f7ac0300bce47dbaeeb23a87b0157cabab0e/nfstream/streamer.py#L276-L303

I am creating streamers with `n_meters=1` specifically because I have only 1 core available (via affinity already set on the parent process) from all CPUs (physically) available on my computing cluster but NFStream is forcing me to `n_meters=0` and trying to autoscale to other CPUs? 
```
WARNING: n_meters set to :0 in order to avoid contention.
```

**Comment by aouinizied** [comment]
@smith558 Indeed this part of the code need a serious rework. Will have a look at it.

**Comment by smith558** [comment]
Currently, the CPU scaling seems to be completely broken for me. I am running the version built from the source.  Even if setting `n_meters=3` the utilization shows only single core is used.
![image](https://user-images.githubusercontent.com/44023416/225627561-d4fdcf05-0cc5-4ea7-95b2-e4ffac318f42.png)

**Comment by smith558** [comment]
After looking into this closer I found out the issue with `n_meters` was on my side in my application logic, however, I am still unable to get above single core utilisation using `n_meters > 1`

```
CPUs: 4
n_meters: 3
CPUs available:  [6, 7, 8, 9]
setting process to: [8, 9]
CPUs available:  [6, 7, 8, 9]
setting process to: [8, 9]
CPUs available:  [6, 7, 8, 9]
setting process to: [6, 7]
CPUs available:  [6, 7, 8, 9]
setting process to: [6, 7]
```

![image](https://user-images.githubusercontent.com/44023416/225665112-b94e0ed1-bdf2-4f81-9def-b25899287c08.png)


Below is the test `n_meters=3` on my desktop PC with AMD (what it should look like):
```
CPUs: 16
n_meters: 3
CPUs available:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
setting process to: [2, 3]
CPUs available:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
setting process to: [4, 5]
CPUs available:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
setting process to: [6, 7]
```
![image](https://user-images.githubusercontent.com/44023416/225666114-cf669687-aa81-43ee-ab9c-23f668883982.png)
